hawkular/client.py

```diff
@@ -0,0 +1,178 @@
+"""
+from __future__ import unicode_literals
+
+import codecs
+import time
+import collections
+import base64
+import ssl
+
+try:
+    import simplejson as json
+except ImportError:
+    import json
+
+try:
+    # Python 3
+    from urllib.request import Request, urlopen, build_opener, install_opener, HTTPErrorProcessor
+    from urllib.error import HTTPError, URLError
+    from urllib.parse import quote, urlencode
+except ImportError:
+    # Fall back to Python 2's urllib2
+    from urllib2 import Request, urlopen, URLError, HTTPError, HTTPErrorProcessor, build_opener, install_opener
+    from urllib import quote, urlencode
+
+
+class HawkularHTTPErrorProcessor(HTTPErrorProcessor):
+    """
+    Hawkular-Metrics uses http codes 201, 204
+    """
+
+    def http_response(self, request, response):
+        if response.code in [200, 201, 204]:
+            return response
+        return HTTPErrorProcessor.http_response(self, request, response)
+
+    https_response = http_response
+
+class HawkularBaseClient:
+    """
+    Creates new client for Hawkular-Metrics. As tenant_id, give intended tenant_id, even if it's not
+    created yet. To change the instance's tenant_id, use tenant(tenant_id) method
+    """
+    def __init__(self,
+                 tenant_id,
+                 host='localhost',
+                 port=8080,
+                 path=None,
+                 scheme='http',
+                 cafile=None,
+                 context=None,
+                 token=None,
+                 username=None,
+                 password=None):
+        """
+        A new instance of HawkularMetricsClient is created with the following defaults:
+
+        host = localhost
+        port = 8081
+        path = hawkular-metrics
+        scheme = http
+        cafile = None
+
+        The url that is called by the client is:
+
+        {scheme}://{host}:{port}/{2}/
+        """
+        self.tenant_id = tenant_id
+        self.host = host
+        self.port = port
+        self.path = path
+        self.cafile = cafile
+        self.scheme = scheme
+        self.context = context
+        self.token = token
+        self.username = username
+        self.password = password
+
+        opener = build_opener(HawkularHTTPErrorProcessor())
+        install_opener(opener)
+
+        if path is None:
+            class_name = self.__class__.__name__
+            path_components = ''.join(["_" + c.lower() if c.isupper() else c for c in class_name]).strip('_')
+            path_components.pop()
+            self.path = '/'.join(path_components);
+        else:
+            self.path = path
+
+    def _get_base_url(self):
+        return "{0}://{1}:{2}/{3}/".format(self.scheme, self.host, str(self.port), self.path)
+
+    def _get_status_url(self):
+        return self._get_base_url() + 'status'
+
+    def tenant(self, tenant_id):
+        self.tenant_id = tenant_id
+
+    def _http(self, url, method, data=None, decoder=None):
+        res = None
+
+        try:
```

Comment:
@rubenvp8510 I understand the purpose of this `try/except` but the block can't be so large. E.g. why do you catch exception also for the creation of the `Request`? (it's all client-side issue, you shouldn't protect against that... and it will be extremely hard to debug issues).

So my suggestion here is to restrict the scope of this `try/except` only to the tiny part of the http communication with Hawkular.